### DISCUSSION TITLE
How to Go to Definition with specified src if I have a cloned repo in the workspace and a pip package at the same time?

### DISCUSSION THREAD

**Discussion by bruce2233** [microsoft/pylance-release#5147] [Q&A] [answered]
For example, I cloned `transformers` package in the vscode workspace  and also pip install it. When I use Go to Definition. It always jumps to the cloned one. How to choose or configure the specified src to jump to?
![image](https://github.com/microsoft/pylance-release/assets/54824693/70f5c530-0a3b-4d95-9091-487c19edee52)

**Comment by rchiodo** [comment, marked as answer]
What you're really asking is the priority of different paths for resolving where an import comes from. The source for this actually is actually open sourced:
https://github.com/microsoft/pyright/blob/60e4050e0a48a47c6a265f33b5d5084c095f8fd5/packages/pyright-internal/src/analyzer/importResolver.ts#L1550

The order it searches looks like:
- [stubPath](https://microsoft.github.io/pyright/#/configuration?id=main-configuration-options)
- root folder (or workspace folder)
- [extraPaths](https://microsoft.github.io/pyright/#/configuration?id=main-configuration-options)
- site-packages (third party)
- then calls into Pylance which adds the bundled stubs (which are treated as third party).

The [_pickBestImport](https://github.com/microsoft/pyright/blob/60e4050e0a48a47c6a265f33b5d5084c095f8fd5/packages/pyright-internal/src/analyzer/importResolver.ts#L1728) then picks the paths that are considered local over 3rd party. 

Sorry, but I don't think it's possible to get it to pick the pip installed version. In that algorithm above, I believe it's finding it like so:

- stubPath = Not found
- root folder = Found and marked as a `local` import
- extraPaths = Not found
- site-packages = Found and marked as a `3rd party import`
- Pylance bundled = Not found.

The version in the `root` folder (since it's local) is chosen over the 3rd party version. Adding the site-packages folder to the `extraPaths` would mark it as local, but it wouldn't trump the version found in the workspace.

**Reply by bruce2233** [reply]
Thanks a lot.
